FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -0,0 +1,233 @@
+    val keysStorage = KeysStorageImpl.getInstance(context)
+    val existingKeyEntities = keysStorage.getRawKeys()
+    val existingPgpKeyDetailsList = keysStorage.getPgpKeyDetailsList()
+    var passphrase: Passphrase? = null
+
+    for (fetchedPgpKeyDetails in fetchedPgpKeyDetailsList) {
+      if (!fetchedPgpKeyDetails.usableForEncryption) {
+        //ask Tom
+        //we skip a key that is not usable for encryption
+        continue
+      }
+
+      val existingPgpKeyDetails =
+        existingPgpKeyDetailsList.firstOrNull { it.fingerprint == fetchedPgpKeyDetails.fingerprint }
+
+      if (existingPgpKeyDetails != null) {
+        if (fetchedPgpKeyDetails.isNewerThan(existingPgpKeyDetails)) {
+          if (passphrase == null) {
+            passphrase = getUsablePassphraseFromCache()
+          }
+
+          val existingKeyEntity = existingKeyEntities.first {
+            it.fingerprint == existingPgpKeyDetails.fingerprint
+          }
+
+          val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
+          roomDatabase.keysDao().updateSuspend(
+            existingKeyEntity.copy(
+              privateKey = safeVersionOfPrvKey,
+              publicKey = fetchedPgpKeyDetails.publicKey.toByteArray()
+            )
+          )
+        }
+      } else {
+        if (passphrase == null) {
+          passphrase = getUsablePassphraseFromCache()
+        }
+
+        val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
+        val keyEntity = fetchedPgpKeyDetails.toKeyEntity(activeAccount).copy(
+          privateKey = safeVersionOfPrvKey,
+          storedPassphrase = null
+        )
+        roomDatabase.keysDao().insertSuspend(keyEntity)
+      }
+    }
+  }
+
+  /**
+   * We receive a private key from EKM in decrypted format. Before saving it to the local database
+   * we have to protect it with a provided [Passphrase] and encrypt it with [AndroidKeyStore]
+   */
+  private suspend fun protectAndEncryptInternally(
+    passphrase: Passphrase,
+    pgpKeyDetails: PgpKeyDetails
+  ): ByteArray = withContext(Dispatchers.IO) {
+    val protectedPrvKey = PgpKey.encryptKeySuspend(
+      armored = requireNotNull(pgpKeyDetails.privateKey),
+      passphrase = passphrase
+    )
+    val encryptedPrvKeyInternally = KeyStoreCryptoManager.encryptSuspend(protectedPrvKey)
+    return@withContext encryptedPrvKeyInternally.toByteArray()
+  }
+
+  /**
+   * Get the first non-null pass phrase for any of the existing keys already in storage
+   * or throw [EmptyPassphraseException] if it is not exist.
+   */
+  private fun getUsablePassphraseFromCache(): Passphrase {
+    val context: Context = getApplication()
+    val keysStorage = KeysStorageImpl.getInstance(context)
+    val fingerprintOfKeyWithNotEmptyPassphrase = keysStorage.getRawKeys()
+      .map { it.fingerprint }
+      .firstOrNull {
+        keysStorage.getPassphraseByFingerprint(it)?.isEmpty == false
```

Comment:
Is this recognizing pass phrases from both storage and RAM?

Reply:
To be clear, we have all pass phrases only in RAM (if we are talking about unprotected pass phrases). 

- `PassphraseType.DATABASE` -> can be stored in the local database (encrypted via AndroidKeyStore) and in RAM(as is)
- `PassphraseType.RAM` -> can be stored only in RAM(as is)

> Is this recognizing pass phrases from both storage and RAM?

Yes. When we fetch a private key from the local database, decrypt it(via AndroidKeyStore) and store it in RAM we do the same things for pass phrases.

https://github.com/FlowCrypt/flowcrypt-android/blob/8459d4ff0ef06cffad2b0a8671baf96611b82855/FlowCrypt/src/main/java/com/flowcrypt/email/security/KeysStorageImpl.kt#L49

`passPhraseMap` contains all available pass phrases.